src/Infrastructure/LeanCode.Kratos/KratosAuthenticationHandler.cs

```diff
@@ -0,0 +1,155 @@
+
+namespace LeanCode.Kratos;
+
+public class KratosAuthenticationHandler<TOptions> : AuthenticationHandler<TOptions>
+    where TOptions : KratosAuthenticationOptions, new()
+{
+    private static string ToUnixTimeSecondsString(DateTime value) =>
+        new DateTimeOffset(value).ToUnixTimeSeconds().ToString(CultureInfo.InvariantCulture);
+
+    private static ImmutableDictionary<T, string> ExtractEnumNames<T>()
+        where T : struct, Enum
+    {
+        return Enum.GetValues<T>()
+            .Distinct()
+            .Select(
+                v =>
+                    (
+                        Value: v,
+                        Name: (
+                            typeof(T)
+                                .GetField(v.ToString(), BindingFlags.Public | BindingFlags.Static)
+                                ?.GetCustomAttribute(typeof(EnumMemberAttribute)) as EnumMemberAttribute
+                        )?.Value
+                    )
+            )
+            .Where(t => t.Name is not null)
+            .ToImmutableDictionary(t => t.Value, t => t.Name!);
+    }
+
+    private static readonly ImmutableDictionary<MethodEnum, string> AuthenticationMethods =
+        ExtractEnumNames<MethodEnum>();
+
+    private static readonly ImmutableDictionary<KratosAuthenticatorAssuranceLevel, string> AssuranceLevels =
+        ExtractEnumNames<KratosAuthenticatorAssuranceLevel>();
+
+    private readonly IFrontendApi api;
+
+    public KratosAuthenticationHandler(
+        IOptionsMonitor<TOptions> options,
+        ILoggerFactory logger,
+        UrlEncoder encoder,
+        ISystemClock clock,
+        IFrontendApi api
+    )
+        : base(options, logger, encoder, clock)
+    {
+        this.api = api;
+    }
+
+    [SuppressMessage("?", "CA1031", Justification = "Exception is returned to the caller, wrapped in Fail result.")]
+    protected override async Task<AuthenticateResult> HandleAuthenticateAsync()
+    {
+        try
+        {
+            KratosSession? session = null;
+
+            if (Request.Headers.TryGetValue("X-Session-Token", out var token) && !string.IsNullOrEmpty(token))
+            {
+                session = await api.ToSessionAsync(xSessionToken: token);
+            }
+            else if (
+                AuthenticationHeaderValue.TryParse(Request.Headers.Authorization, out var value)
+                && string.Equals(value.Scheme, "bearer", StringComparison.OrdinalIgnoreCase)
+                && !string.IsNullOrEmpty(value.Parameter)
+            )
+            {
+                session = await api.ToSessionAsync(xSessionToken: value.Parameter);
+            }
+            else if (
+                Request.Cookies.TryGetValue(Options.SessionCookieName, out var cookie) && !string.IsNullOrEmpty(cookie)
+            )
+            {
+                session = await api.ToSessionAsync(cookie: $"{Options.SessionCookieName}={cookie}");
+            }
```

Comment:
Let's do a quick refactoring of this method - extract this block into a separate method. Will be easier to read, plus you would be able to get rid of the dangling variable.